feat: add `ndarray/find`

gururaj1512 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: passed
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: passed
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: passed
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: passed ---

Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/70.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds implementation for `ndarray/find`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/70.
-   Progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add ndarray/find

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
@gururaj1512 Before you start doing too many of things in https://github.com/stdlib-js/stdlib/issues/2656 and us needing to go back and fix things, I'd put these on pause. As mentioned in https://github.com/stdlib-js/stdlib/issues/2656, the issue is that `find` (and `every`) are reductions, and thus should support specifying axes over which to perform the reductions. And furthermore, the implementation proposed in this PR may work for the flattened case; however, it won't work for the axes case. And lastly, we should always return ndarrays, not scalars. In short, there is a bit of R&D which needs to be done in order for these to move forward.

kgryte commented:
One other thing: each top-level ndarray implementation should be a wrapper around a "base" implementation. Notice how in `base/for-each`, we write kernels for efficient iteration: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ndarray/base/for-each/lib. I would expect something similar for reductions.

kgryte commented:
So before top-level implementations, we need to create the "base" implementations.

gururaj1512 commented:
> @gururaj1512 Before you start doing too many of things in #2656 and us needing to go back and fix things, I'd put these on pause. As mentioned in #2656, the issue is that `find` (and `every`) are reductions, and thus should support specifying axes over which to perform the reductions. And furthermore, the implementation proposed in this PR may work for the flattened case; however, it won't work for the axes case. And lastly, we should always return ndarrays, not scalars. In short, there is a bit of R&D which needs to be done in order for these to move forward.

Yaa I had similar doubts while creating this PR. Just created this quick PR to discuss about it.

gururaj1512 commented:
@kgryte, Should we pause this, or is there something I can work on? I would love to work on issues related to this.

kgryte commented:
@gururaj1512 Yes, let's pause. I pinged you on the tracking issue.

kgryte commented:
@headlessNode Once https://github.com/stdlib-js/stdlib/pull/7426 is merged, would you mind taking this PR over, refactoring, and basing on `ndarray/base/find`?

kgryte commented:
@headlessNode Now that #7426 has been merged, this is now ready for you.

headlessNode commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray/find

headlessNode commented:
/stdlib help

headlessNode commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

headlessNode commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray/find

headlessNode committed:
feat: add ndarray/find

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+// returns <ndarray>
+
+var v = ndarray2array( out );
+// returns [ NaN, 2.0, NaN, 4.0 ]
```
```suggestion
// returns [ [ NaN, 2.0 ], [ NaN, 4.0 ] ]
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+// returns <ndarray>
+
+// Create an output ndarray:
+var y = empty( [ 4 ], {
```
```suggestion
var y = empty( [ 2, 2 ], {
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+// returns true
+
+var v = ndarray2array( y );
+// returns [ NaN, 2.0, NaN, 4.0 ]
```
```suggestion
// returns [ [ NaN, 2.0 ], [ NaN, 4.0 ] ]
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+console.log( ndarray2array( x ) );
+
+var y = find( x, isEven );
+console.log( y );
```
```suggestion
console.log( y.get() );
```

headlessNode replied:
```suggestion
console.log( y.get() );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/examples/index.js:
```diff
@@ -0,0 +1,37 @@
+console.log( ndarray2array( x ) );
+
+var y = find( x, isEven );
+console.log( y );
```
```suggestion
console.log( y.get() );
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+}
+
+// Create an input ndarray:
+var x = array( [ [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ], [ [ 0.0, 6.0 ], [ 7.0, 8.0 ] ] ] );
```
```suggestion
var x = array( [ [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ], [ 7.0, 8.0 ] ] ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+}
+
+// Create an input ndarray:
+var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );
```
```suggestion
var x = array( [ [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ], [ 7.0, 8.0 ] ] ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+
+// Create an output ndarray:
+var y = empty( [ 2, 2 ], {
+    'dtype': 'bool'
```
```suggestion
    'dtype': 'x.dtype'
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/find/README.md:
```diff
@@ -0,0 +1,338 @@
+
+// Create an output ndarray:
+var y = empty( [ 2, 2 ], {
+    'dtype': 'x.dtype'
```
```suggestion
    'dtype': x.dtype
```

headlessNode committed:
fix: lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/main.js:
```diff
@@ -0,0 +1,224 @@
+		throw new TypeError( format( 'invalid argument. First argument must be an ndarray-like object. Value: `%s`.', x ) );
+	}
+
+	if ( nargs === 2 ) {
```
```suggestion
	if ( nargs <= 2 ) {
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/main.js:
```diff
@@ -0,0 +1,224 @@
+				throw new TypeError( format( 'invalid argument. Third argument must be a function. Value: `%s`.', predicate ) );
+			}
+		}
+	} else if ( nargs === 5 ) {
```
```suggestion
	} else { // nargs >= 5
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/main.js:
```diff
@@ -0,0 +1,224 @@
+		}
+	} else {
+		throw new Error( 'invalid number of arguments.' );
+	}
```
```suggestion
	}
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/main.js:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	if ( isndarrayLike( sentinelValue ) ) {
+		try {
```
Is it necessary to wrap this in a `try/catch`? Why not just allow `maybeBroadcastArray` to raise an exception?

kgryte replied:
It is possible that we have this pattern elsewhere, but I think we also just allow `maybeBroadcastArray` to throw.

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/assign.js:
```diff
@@ -0,0 +1,207 @@
+		throw new TypeError( format( 'invalid argument. Second argument must be an ndarray-like object. Value: `%s`.', y ) );
+	}
+
+	if ( nargs === 3 ) {
```
```suggestion
	if ( nargs <= 3 ) {
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/assign.js:
```diff
@@ -0,0 +1,207 @@
+				throw new TypeError( format( 'invalid argument. Fourth argument must be a function. Value: `%s`.', predicate ) );
+			}
+		}
+	} else if ( nargs === 6 ) {
```
```suggestion
	} else { // nargs >= 6
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/assign.js:
```diff
@@ -0,0 +1,207 @@
+		}
+	} else {
+		throw new Error( 'invalid number of arguments.' );
+	}
```
```suggestion
	}
```

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/assign.js:
```diff
@@ -0,0 +1,207 @@
+* Finds the first elements which pass a test implemented by a predicate function along one or more ndarray dimensions and assigns results to a provided output ndarray.
+*
+* @param {ndarray} x - input ndarray
+* @param {ndarray} y - output ndarray
```
Let's rename `y` to `out`.

kgryte commented on lib/node_modules/@stdlib/ndarray/find/lib/assign.js:
```diff
@@ -0,0 +1,207 @@
+*
+* @param {ndarray} x - input ndarray
+* @param {ndarray} y - output ndarray
+* @param {*|ndarray} [sentinelValue] - sentinel value
```
Let's move `sentinelValue` to the second argument before the output ndarray. This would be similar to `ndarray/some-by` and `blas/ext/index-of`.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte ping

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (dismissed):
LGTM after clean-up.

kgryte committed:
docs: fix examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---
